lib/src/migrators/module.dart

```diff
@@ -46,54 +35,55 @@ class ModuleMigrator extends MigratorBase {
   ///
-  /// If [_migrateDependencies] is false, the migrator will still be run on
+  /// If [migrateDependencies] is false, the migrator will still be run on
   /// dependencies, but they will be excluded from the resulting map.
-  @override
-  p.PathMap<String> migrateFile(String entrypoint) {
-    var migration = _migrateStylesheet(entrypoint, Directory.current.path);
-    var results = p.PathMap<String>();
-    addMigration(StylesheetMigration migration) {
-      var migrated = migration.migratedContents;
-      if (migrated != migration.contents) {
-        results[migration.path] = migrated;
-      }
-    }
-
-    if (_migrateDependencies) {
-      _migrations.values.forEach(addMigration);
-    } else {
-      addMigration(migration);
+  void migrateFile(String entrypoint) {
+    _ModuleMigrationVisitor(this, entrypoint).run();
+    if (!migrateDependencies) {
+      migrated.removeWhere((path, contents) => path != entrypoint);
     }
-    return results;
   }
+}
 
-  /// Migrates the stylesheet at [path] if it hasn't already been migrated and
-  /// returns the StylesheetMigration instance for it regardless.
-  StylesheetMigration _migrateStylesheet(String path, String directory) {
-    path = canonicalizePath(p.join(directory, path));
-    return _migrations.putIfAbsent(path, () {
-      var migration = StylesheetMigration(path);
-      _activeMigrations.add(migration);
-      visitStylesheet(migration.stylesheet);
-      _activeMigrations.remove(migration);
-      return migration;
-    });
+class _ModuleMigrationVisitor extends MigrationVisitor {
+  final ModuleMigrator migrator;
+  final String path;
+
+  _ModuleMigrationVisitor(this.migrator, this.path);
+
+  _ModuleMigrationVisitor newInstance(String newPath) =>
+      _ModuleMigrationVisitor(migrator, newPath);
+
+  /// Namespaces of modules used in this stylesheet.
+  final namespaces = p.PathMap<String>();
+
+  /// Set of additional use rules necessary for referencing members of
+  /// implicit dependencies / built-in modules.
+  ///
+  /// This set contains the path provided in the use rule, not the canonical
+  /// path (e.g. "a" rather than "dir/a.scss").
+  final additionalUseRules = Set<String>();
+
+  /// Global variables declared with !default that could be configured.
+  final configurableVariables = normalizedSet();
+
+  /// Local variables, mixins, and functions for this migration.
+  ///
+  /// When at the top level of the stylesheet, this will be null.
+  LocalScope localScope;
+
+  /// Returns the migrated contents of this stylesheet, based on [patches] and
+  /// [additionalUseRules], or null if the stylesheet does not change.
+  @override
+  String getMigratedContents() {
+    if (patches.isEmpty && additionalUseRules.isEmpty) return null;
+    var semicolon = syntax == Syntax.sass ? "" : ";";
+    var uses = additionalUseRules.map((use) => '@use "$use"$semicolon\n');
+    var contents = Patch.applyAll(stylesheet.span.file, patches);
```

Comment:
Is there a reason to avoid `super.getMigratedContents()` here?

Reply:
I was thinking of the potential case where there are additional use rules, but no patches, but I forgot that if a use rule is added, there would be at least one patch to add a namespace to a reference from that file.